File: NSL.js
function findNearestSmallestElement(arr) {
  let vector = [];
  let stack = [];

  for (let i = arr.length - 1; i >= 0; i--) {
    while (stack.length > 0 && arr[i] <= stack[stack.length - 1]) {
      stack.pop();
    }

    vector[i] = stack.length === 0 ? -1 : stack[stack.length - 1];

    stack.push(arr[i]);
  }

  return vector;
}

console.log(findNearestSmallestElement([4, 5, 2, 10, 8]));

File: NSL.py
# def NearestSmallestElement(arr): Brute Force
#     n = len(arr)
#     for i in range(n):
#         for j in range(i + 1, n):
#             if arr[i] > arr[j]:
#                 arr[i], arr[j] = arr[j], arr[i]
#     return arr

## Optimatized

def NearestSmallestElement(arr):
    vector = []
    stack = []

    for i in range(len(arr)):
        if len(stack) == 0:
            vector.append(-1)
        elif len(stack) > 0 and stack[0] < arr[i]:
            vector.append(stack[0])
        else:
            while (len(stack) > 0 and stack[0] >= arr[i]):
                stack.pop(0)
            if len(stack) == 0:
                vector.append(-1)
            else:
                vector.append(stack[0])

        stack.append(arr[i])

    return vector

print(NearestSmallestElement([4, 5, 2, 10, 8]))